### ISSUE TITLE
langmap produces unexpected output in select mode

### ISSUE THREAD

**Issue by Anstow** [vim/vim#572]
When langmap is set, the first character entered is translated using langmap before being entered.

For example, if `langmap=ab` is set and vim is in select mode then when `aaa` is entered the result is

```
baa
```

whereas the result I expect is

```
aaa
```

It doesn't seem to make a difference whether `langnoremap` is set or not.

**Comment by chrisbra** [comment]
I see the problem. I think this fixes the problem:

``` patch
diff --git a/src/getchar.c b/src/getchar.c
index 70829e1..b06c38d 100644
--- a/src/getchar.c
+++ b/src/getchar.c
@@ -2149,7 +2149,9 @@ vgetorpeek(advance)
                        else
                        {
                            LANGMAP_ADJUST(c1,
-                                          (State & (CMDLINE | INSERT)) == 0);
+                                          ((State & (CMDLINE | INSERT) &&
+                                            get_real_state() != SELECTMODE)
+                                           ) == 0);
                            nolmaplen = 0;
                        }
 #endif
```

**Comment by chrisbra** [comment]
(sorry for wrong content, i corrected the previous comment)

**Comment by chrisbra** [comment]
patch should look like this (repost for vim-dev ml)

``` patch
diff --git a/src/getchar.c b/src/getchar.c
index 70829e1..b06c38d 100644
--- a/src/getchar.c
+++ b/src/getchar.c
@@ -2149,7 +2149,9 @@ vgetorpeek(advance)
                        else
                        {
                            LANGMAP_ADJUST(c1,
-                                          (State & (CMDLINE | INSERT)) == 0);
+                                          ((State & (CMDLINE | INSERT) &&
+                                            get_real_state() != SELECTMODE)
+                                           ) == 0);
                            nolmaplen = 0;
                        }
 #endif
```

**Comment by Anstow** [comment]
Thanks for the quick patch. 

Could you explain to me what happens if there are only non-printable characters in the langmap? (Or some combination of non-printable characters and printable characters?)

**Comment by chrisbra** [comment]
Hi David!

On Mi, 20 Jan 2016, David Watson wrote:

> Thanks for the quick patch. 
> 
> Could you explain to me what happens if there are only non-printable characters in the langmap? (Or some combination of non-printable characters and printable characters?)

I am sorry, I don't follow.

Best,

## Christian

Den Geschmack kann man nicht am Mittelgut bilden, sondern nur am
Allervorzüglichsten.
        -- Johann Peter Eckermann

**Comment by Anstow** [comment]
Sorry it appears langmap doesn't work on nonprintable keys.

I was thinking about something along the lines of
    noremap <c-n> diw
    set langmap=<c-n><c-j>
Then (assuming langmap worked in this way) one would expect <c-j> to delete a word. Whereas with the above change I think <c-n> would still delete a word in select mode.

**Comment by Anstow** [comment]
It appears that the markdown format stripped out some of my example, I have corrected the previous comment. (Corrected here for vim-dev ml.)
Sorry it appears langmap doesn't work on nonprintable keys.

I was thinking about something along the lines of
    noremap <c-n> diw
    set langmap=<c-n><c-j>
Then (assuming langmap worked in this way) one would expect <c-j> to delete a word. Whereas with the above change I think <c-n> would still delete a word in select mode.

**Comment by vim-ml** [comment]
Hi David!

On Mi, 20 Jan 2016, David Watson wrote:

> Sorry it appears langmap doesn't work on nonprintable keys.
> 
> I was thinking about something along the lines of
>     noremap <c-n> diw
>     set langmap=<c-n><c-j>

I am not sure. Try to enter the keys literally, e.g. press <c-v><c-w><c-v><c-j>

> Then (assuming langmap worked in this way) one would expect <c-j> to delete a word. Whereas with the above change I think <c-n> would still delete a word in select mode. 

mappings currently don't apply to resolved langmaps. That is a bug or 
not and has been discussed previously on vim-dev

Best,

## Christian

Schlagzeile einer italienischen Zeitung: "Die Bibliothek von Totti ist in
Flammen aufgegangen, sie enthielt zwei Bücher."
Totti ist verzweifelt: "Verdammt, das zweite hatte ich noch gar nicht ausgemalt!"

**Comment by brammool** [comment]
Christian Brabandt wrote:

> patch should look like this (repost for vim-dev ml)
> 
> ``` patch
> diff --git a/src/getchar.c b/src/getchar.c
> index 70829e1..b06c38d 100644
> --- a/src/getchar.c
> +++ b/src/getchar.c
> @@ -2149,7 +2149,9 @@ vgetorpeek(advance)
>                         else
>                         {
>                             LANGMAP_ADJUST(c1,
> -                                          (State & (CMDLINE | INSERT)) == 0);
> +                                          ((State & (CMDLINE | INSERT) &&
> +                                            get_real_state() != SELECTMODE)
> +                                           ) == 0);
>                             nolmaplen = 0;
>                         }
>  #endif
> ```

Thanks!

Bonus points for a test?

## 

hundred-and-one symptoms of being an internet addict:
24. You realize there is not a sound in the house and you have no idea where
    your children are.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\
\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by vim-ml** [comment]
Hi Bram!

On Mi, 20 Jan 2016, Bram Moolenaar wrote:

> Christian Brabandt wrote:
> 
> > patch should look like this (repost for vim-dev ml)
> > 
> > ``` patch
> > diff --git a/src/getchar.c b/src/getchar.c
> > index 70829e1..b06c38d 100644
> > --- a/src/getchar.c
> > +++ b/src/getchar.c
> > @@ -2149,7 +2149,9 @@ vgetorpeek(advance)
> >                         else
> >                         {
> >                             LANGMAP_ADJUST(c1,
> > -                                          (State & (CMDLINE | INSERT)) == 0);
> > +                                          ((State & (CMDLINE | INSERT) &&
> > +                                            get_real_state() != SELECTMODE)
> > +                                           ) == 0);
> >                             nolmaplen = 0;
> >                         }
> >  #endif
> > ```
> 
> Thanks!
> 
> Bonus points for a test?

Sure. But not today.

Best,

## Christian

Es ist eine besonders hübsche Kunst, sich in Dinge einweihen zu
lassen, über die man Bescheid weiß, und zwar von Leuten, die nicht
Bescheid wissen.
        -- Alain Peyrefitte

**Comment by vim-ml** [comment]
Hi Bram!

On Mi, 20 Jan 2016, Bram Moolenaar wrote:

> Christian Brabandt wrote:
> 
> > patch should look like this (repost for vim-dev ml)
> > 
> > ``` patch
> > diff --git a/src/getchar.c b/src/getchar.c
> > index 70829e1..b06c38d 100644
> > --- a/src/getchar.c
> > +++ b/src/getchar.c
> > @@ -2149,7 +2149,9 @@ vgetorpeek(advance)
> >                         else
> >                         {
> >                             LANGMAP_ADJUST(c1,
> > -                                          (State & (CMDLINE | INSERT)) == 0);
> > +                                          ((State & (CMDLINE | INSERT) &&
> > +                                            get_real_state() != SELECTMODE)
> > +                                           ) == 0);
> >                             nolmaplen = 0;
> >                         }
> >  #endif
> > ```
> 
> Thanks!
> 
> Bonus points for a test?

Hm, that was wrong. Not sure, why I thought this fixes it. Good you 
asked for a test.

Updated patch attached.

Best,

## Christian

Im Leben lernt der Mensch zuerst das Gehen und Sprechen. Später lernt
er dann stillzusitzen und den Mund zu halten.
        -- Marcel Pagnol

commit e0f6e7ff3d7137bde651bb19f22e82ff2814e2b5
Author: Christian Brabandt cb@256bit.org
Date:   Thu Jan 21 22:38:51 2016 +0100

```
fix issue #572
```

diff --git a/src/normal.c b/src/normal.c
index ac1a391..17cb8d6 100644
--- a/src/normal.c
+++ b/src/normal.c
@@ -638,7 +638,7 @@ normal_cmd(oap, toplevel)
      \* Get the command character from the user.
      */
     c = safe_vgetc();
-    LANGMAP_ADJUST(c, TRUE);
-    LANGMAP_ADJUST(c, get_real_state() != SELECTMODE);
  
   /*
  - If a mapping was started in Visual or Select mode, remember the length
    diff --git a/src/testdir/test_alot.vim b/src/testdir/test_alot.vim
    index e89afb4..705301d 100644
    --- a/src/testdir/test_alot.vim
    +++ b/src/testdir/test_alot.vim
    @@ -5,6 +5,7 @@ source test_backspace_opt.vim
    source test_cursor_func.vim
    source test_delete.vim
    source test_expand.vim
    +source test_langmap.vim
    source test_lispwords.vim
    source test_menu.vim
    source test_searchpos.vim
    diff --git a/src/testdir/test_langmap.vim b/src/testdir/test_langmap.vim
    new file mode 100644
    index 0000000..2c4ea49
    --- /dev/null
    +++ b/src/testdir/test_langmap.vim
    @@ -0,0 +1,11 @@
    +" Tests for 'langmap'
    +
    +func Test_langmap()
-  new
-  $put ='foobar foobar foobar'
-  set langmap=ab
-  :2
-  :norm! gHaaa
-  call assert_equal(getline(1,'$'), ['','aaa'])
  +
  +endfunc

**Comment by vim-ml** [comment]
Christian Brabandt wrote:

> Hi Bram!
> 
> On Mi, 20 Jan 2016, Bram Moolenaar wrote:
> 
> > Christian Brabandt wrote:
> > 
> > > patch should look like this (repost for vim-dev ml)
> > > 
> > > ``` patch
> > > diff --git a/src/getchar.c b/src/getchar.c
> > > index 70829e1..b06c38d 100644
> > > --- a/src/getchar.c
> > > +++ b/src/getchar.c
> > > @@ -2149,7 +2149,9 @@ vgetorpeek(advance)
> > >                         else
> > >                         {
> > >                             LANGMAP_ADJUST(c1,
> > > -                                          (State & (CMDLINE | INSERT)) == 0);
> > > +                                          ((State & (CMDLINE | INSERT) &&
> > > +                                            get_real_state() != SELECTMODE)
> > > +                                           ) == 0);
> > >                             nolmaplen = 0;
> > >                         }
> > >  #endif
> > > ```
> > 
> > Thanks!
> > 
> > Bonus points for a test?
> 
> Hm, that was wrong. Not sure, why I thought this fixes it. Good you 
> asked for a test.
> 
> Updated patch attached.

Thanks.  I had discovered the same thing.  I tried using feedkeys() for
the test, but had to add an argument to make it work.  ":normal!" also
works, as your test shows, but I actually doubt if that is correct, it
should probably not apply 'langmap' there, since these are not typed
keys.

## 

For a moment, nothing happened.
Then, after a second or so, nothing continued to happen.
        -- Douglas Adams, "The Hitchhiker's Guide to the Galaxy"

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\
\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\            help me help AIDS victims -- http://ICCF-Holland.org    ///
